### DISCUSSION TITLE
How to use pyproject.toml for package distribution

### DISCUSSION THREAD

**Discussion by Viatorus** [pyscaffold/pyscaffold#412] [Q&A] [answered]
Currently I do:

```
- pip3 install --upgrade setuptools setuptools_scm wheel twine
- python3 setup.py sdist bdist_wheel
- python3 -m twine upload dist/*
```

In the documentation it is mentioned to use pyproject.toml instead of the above code:

> If you prefer to have legacy builds and get the old behavior, you can remove the pyproject.toml file and run python setup.py bdist_wheel

https://pyscaffold.org/en/latest/faq.html?highlight=pyproject.toml#pyproject-toml

But how?

**Comment by abravalheri** [comment, marked as answer]
Hi @Viatorus, if you are not experiencing any problems with `pyproject.toml`, your current flow is perfectly fine.

The idea behind `pyproject.toml` is that you should specify build dependencies (such as `wheel` and `setuptools_scm`) there ([`pyproject.toml > [build-system] > requires`](https://github.com/pyscaffold/pyscaffold-demo/blob/master/pyproject.toml#L3)).

In v4, [`tox.ini`](https://github.com/pyscaffold/pyscaffold-demo/blob/master/tox.ini#L24) is preconfigured to work with the newest suggested workflow from PyPA that uses `pyproject.toml` and [`build`](https://pypa-build.readthedocs.io/en/latest/). The idea behind this new workflow is that it is generic for any other build tool, but if you are using setuptools (which is what PyScaffold adopts anyway), `python3 setup.py bdist_wheel` should work just fine.

(In a distant future I suppose `setup.py` will be retired, but I think we are still far from that).

**Reply by Viatorus** [reply]
Thank you for your answer and clarification. I will stick to setuptools for now.

**Comment by raphaelquast** [comment]
I just stumbled upon this issue and since it doesn't really give an answer on what to do if I want to stick with a `pyproject.toml` I thought I'll add my 2 cents for future reference... 

In short, the commands you've been searching for are:
 (more details can be found in the [packaging guide](https://packaging.python.org/en/latest/tutorials/packaging-projects/#generating-distribution-archives))

```
python3 -m pip install --upgrade build
python3 -m build
python3 -m twine upload dist/*
```
